### ISSUE TITLE
[FR] gRPC inference server for pyfunc flavor 

### ISSUE THREAD

**Issue by tokoko** [mlflow/mlflow#4190]
## Willingness to contribute
- [ ] Yes. I can contribute this feature independently.
- [x] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary

Current `FlavorBackend` implementation for `pyfunc` model flavor serves REST API endpoint for inference. The proposed feature would add an option to deploy `pyfunc` models as gRPC endpoints as well. I have begun a rough draft of the feature in a separate repository ([mlflow-grpc](https://github.com/tokoko/mlflow-grpc)) but i'm unsure how the feature can be incorporated into the main `mlflow` project.

## Motivation
- What is the use case for this feature?
deployment of mlflow pyfunc models as gRPC endpoints. 
- Why is this use case valuable to support for MLflow users in general?
ability to expose faster gRPC inference endpoints might be crucial in production for many ml use cases.
- Why is this use case valuable to support for your project(s) or organization?
- Why is it currently difficult to achieve this use case? (please be as specific as possible about why related MLflow features and components are insufficient)
It could be implemented by writing a custom Dockerfile and protobuf definitions and keeping them up-to-date with changes in model dependencies and/or input features.

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

## Details

Seems to me, there are several ways to incorporate this feature into the project, but I don't know which if any would be the right approach.

- implement it as part of `mlflow.pyfunc.backend.PyFuncBackend` in already existing serve() and build_docker() functions. they would need additional parameters to determine whether to serve REST or gRPC. `mlflow models` API will need to be changed accordingly. This approach seems too messy.

- implement it as a separate `FlavorBackend`. This seems like a cleaner approach, but there are some problems. `get_flavor_backend` function in `mlflow.models.flavor_backend_registry` right now assumes that there will be at most one `FlavorBackend` per each model `flavor`. if this assumption was dropped, `mlflow models` API would need to be changed to add an additional parameter to specify which `FlavorBackend` should be used. another minor issue with this approach is that it will probably never make sense to have more than one `predict()` implementations per `flavor` and flavor backends would probably need to share it. The way I decided to implement my POC version of the feature is by writing `PyFuncGrpcBackend` that extends `PyFuncBackend` and overrides only `serve` and `build_docker` methods.

- implement it as a plugin. If I'm not mistaken `FlavorBackend` is not pluggable as of now and I'm not sure if `deployments` plugin mechanism is appropriate for this use case. would love to hear your view on this.

**Comment by cagdasbas** [comment]
Hi @tokoko ,
The default implementation of gRPC really slows the things down. For example, I have a model that runs at 80FPS within a loop. If I serve it with gRPC, I can only get ~20fps on the client side. But instead of regular gRPC server, if I use asyncio grpc (check [grpclib](https://github.com/vmagamedov/grpclib) ), I can get ~75fps on the client side. You should check it. The client side isn't changing, you can still use regular gRPC stubs, this change doesn't require a client modification.

**Comment by ulhaqi12** [comment]
Are there any updates on this?
Is anyone working to provide gRPC support for MLFlow?

-Ikram

**Comment by tokoko** [comment]
@ulhaqi12 I don't think so. You should probably try [mlserver](https://mlserver.readthedocs.io/en/latest/examples/mlflow/README.html) mlflow integration. It can deploy mlflow models with both http and grpc versions of [Open Inference Protocol](https://github.com/kserve/open-inference-protocol).
